documentation/man/hmmpgmd.man
.TH "hmmpgmd" 1 "@HMMER_DATE@" "HMMER @HMMER_VERSION@" "HMMER Manual"

.SH NAME
hmmpgmd - daemon for searching a protein query against a protein database 


.SH SYNOPSIS
.B hmmpgmd
.I [options]


.SH DESCRIPTION

.PP
The
.B hmmpgmd 
program is the daemon that we use internally for the hmmer.org web server, 
and essentially stands in front of the protein search programs
.BR phmmer , 
.BR hmmsearch ,
and 
.BR hmmscan . 

To use
.BR hmmpgmd , 
first an instance must be started up as a 
.I master 
server, and provided with at least one of a 
.I "sequence database"
(using the 
.B --seqdb
flag)
and/or an 
.I "HMM database"
(using the
.B --hmmdb
flag). 
A sequence database must be in the hmmpgmd format, which may be
produced using 
.BR esl-reformat .
An HMM database is of the form produced by 
.BR hmmbuild .
The input database(s) will be loaded into memory by the 
master. When the master has finished loading the database(s), it 
prints the line:
"Data loaded into memory. Master is ready."


.PP
Only after master is ready, one or more instances of hmmpgmd may
be started as workers. These workers may be (and typically are) on
different machines from the master, but must have access to the 
same database file(s) provided to the master, with the same path. As 
with the master, each worker loads the database(s) into memory, and 
indicates completion by printing: "Data loaded into memory. Worker is ready."


.PP
The master server and workers are expected to remain running.
One or more clients then connect to the master and submit possibly
many queries. The master distributes the work of a query among the
workers, collects results, and merges them before responding to the
client. Two example client programs are included in the HMMER3.1 src 
directory - the C program
.B hmmc2
and the perl script
.BR hmmpgmd_client_example.pl .
These are intended as examples only, and should be extended as 
necessary to meet your needs. 

.PP
A query is submitted to the master from the client as a character
string. Queries may be the sort that would normally be handled
by 
.B phmmer
(protein sequence vs protein database),
.B hmmsearch
(protein HMM query vs protein database), or
.B hmmscan
(protein query vs protein HMM database).
 

The general form of a client query is to start with a single line
of the form 
.IR "@[options]" , 
followed by multiple lines of text representing either the query HMM 
or fasta-formatted sequence. The final line of each query is the separator 
.IR "//" .


.PP
For example, to perform a 
.B phmmer
type search of a sequence against a sequence database 
file, the first line is of the form 
.IR "@--seqdb 1" ,
then the fasta-formatted query sequence starting with the header line
.IR >sequence-name ,
followed by one or more lines of sequence, and finally the closing
.IR "//" .

.PP
To perform an
.B hmmsearch 
type search, the query sequence is replaced by the full
text of a HMMER-format query HMM. 

.PP
To perform an
.B hmmscan 
type search, the text matches that of the 
.B phmmer
type search, except that the first line changes to 
.IR "@--hmmdb 1" .

.PP
In the hmmpgmd-formatted sequence database file, each sequence
can be associated with one or more sub-databases. The 
.I --seqdb
flag indicates which of these sub-databases will be queried. 
The HMM database format does not support sub-databases.

.PP
The result of each query is an undocumented data structure in 
binary format. In the future the data will be returned in a proper
serialized structure, but for now, it requires meticulous unpacking
within the client. The example clients show how this is done.

 

.SH OPTIONS

.TP
.B -h
Help; print a brief reminder of command line usage and all available
options.



.SH EXPERT OPTIONS

.TP 
.BI --master
Run as the master server.

.TP
.BI --worker " <s>"
Run as a worker, connecting to the master server that is running on IP
address
.IR <s> .

.TP 
.BI --daemon
Run as a daemon using config file: /etc/hmmpgmd.conf

.TP 
.BI --cport " <n>"
Port to use for communication between clients and the master server. 
The default is 51371.

.TP 
.BI --wport " <n>"
Port to use for communication between workers and the master server. 
The default is 51372.

.TP 
.BI --ccncts " <n>"
Maximum number of client connections to accept. The default is 16.

.TP 
.BI --wcncts " <n>"
Maximum number of worker connections to accept. The default is 32.

.TP 
.BI --pid " <f>"
Name of file into which the process id will be written. 

.TP 
.BI --seqdb " <f>"
Name of the file (in hmmpgmd format) containing protein sequences.
The contents of this file will be cached for searches. 

.TP 
.BI --hmmdb " <f>"
Name of the file containing protein HMMs. The contents of this file 
will be cached for searches.

.TP 
.BI --cpu " <n>"
Number of parallel threads to use (for 
.I --worker
).


.SH SEE ALSO 

See 
.B hmmer(1)
for a master man page with a list of all the individual man pages
for programs in the HMMER package.

.PP
For complete documentation, see the user guide that came with your
HMMER distribution (Userguide.pdf); or see the HMMER web page
(@HMMER_URL@).



.SH COPYRIGHT

.nf
@HMMER_COPYRIGHT@
@HMMER_LICENSE@
.fi

For additional information on copyright and licensing, see the file
called COPYRIGHT in your HMMER source distribution, or see the HMMER
web page 
(@HMMER_URL@).


.SH AUTHOR

.nf
Eddy/Rivas Laboratory
Janelia Farm Research Campus
19700 Helix Drive
Ashburn VA 20147 USA
http://eddylab.org
.fi





src/daemon/hmmpgmd.c
/* HMMER search daemon
 */
#include "p7_config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>

#ifndef HMMER_THREADS
#error "Program requires pthreads be enabled."
#endif /*HMMER_THREADS*/

#include "easel.h"
#include "esl_getopts.h"

#include "hmmer.h"

#define CONF_FILE "/etc/hmmpgmd.conf"

static ESL_OPTIONS cmdlineOpts[] = {
  /* name           type         default      env   range           toggles  reqs   incomp           help                                                     docgroup */
  { "-h",           eslARG_NONE,   FALSE,     NULL, NULL,           NULL,  NULL,  NULL,            "show brief help on version and usage",                         1 },
  { "--master",     eslARG_NONE,    NULL,     NULL, NULL,           NULL,  NULL,  "--worker",      "run program as the master server",                            12 },
  { "--worker",     eslARG_STRING,  NULL,     NULL, NULL,           NULL,  NULL,  "--master",      "run program as a worker with server at <s>",                  12 },
  { "--cport",      eslARG_INT,     "51371",  NULL, "49151<n<65536",NULL,  NULL,  "--worker",      "port to use for client/server communication",                 12 },
  { "--wport",      eslARG_INT,     "51372",  NULL, "49151<n<65536",NULL,  NULL,  NULL,            "port to use for server/worker communication",                 12 },
  { "--ccncts",     eslARG_INT,     "16",     NULL, "n>0",          NULL,  NULL,  "--worker",      "maximum number of client side connections to accept",         12 },
  { "--wcncts",     eslARG_INT,     "32",     NULL, "n>0",          NULL,  NULL,  "--worker",      "maximum number of worker side connections to accept",         12 },
  { "--pid",        eslARG_OUTFILE, NULL,     NULL, NULL,           NULL,  NULL,  NULL,            "file to write process id to",                                 12 },
  { "--daemon",     eslARG_NONE,    NULL,     NULL, NULL,           NULL,  NULL,  NULL,            "run as a daemon using config file: /etc/hmmpgmd.conf",        12 },
  { "--seqdb",      eslARG_INFILE,  NULL,     NULL, NULL,           NULL,  NULL,  "--worker",      "protein database to cache for searches",                      12 },
  { "--hmmdb",      eslARG_INFILE,  NULL,     NULL, NULL,           NULL,  NULL,  "--worker",      "hmm database to cache for searches",                          12 },
  { "--cpu",        eslARG_INT,     NULL,"HMMER_NCPU","n>0",        NULL,  NULL,  "--master",      "number of parallel CPU workers to use for multithreads",      12 },
  {  0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
};

static char usage[]  = "[options]";
static char banner[] = "search a query against a database";


typedef void sig_func(int);

/* MSF called this <signal()>. That conflicts with system's <signal()>. 
 * Unclear what he was doing. This appears to duplicated functionality
 * of the system's <signal()>. Revisit later. For now, silence compiler
 * warning by renaming the function.
 */
sig_func *
our_signal(int signo, sig_func *fn)
{
  struct sigaction act;
  struct sigaction oact;

  act.sa_handler = fn;
  sigemptyset(&act.sa_mask);
  act.sa_flags = 0;
  if (signo == SIGALRM) {
#ifdef SA_INTERRUPT
    act.sa_flags |= SA_INTERRUPT;  /* SunOS 4.x */
#endif
  } else {
#ifdef SA_RESTART
    act.sa_flags |= SA_RESTART;  /* SVR4, 4.4BSD */
#endif
  }
  if (sigaction(signo, &act, &oact) < 0) {
    return SIG_ERR;
  }

  return oact.sa_handler;
}

/* write_pid()
 * Log the process id to a file.
 */
static void
write_pid(ESL_GETOPTS *go)
{
  char   *pid_file = esl_opt_GetString(go, "--pid");
  FILE   *fp       = fopen(pid_file, "w");

  if (!fp) p7_Fail("Unable to open PID file %s for writing.", pid_file);
  fprintf(fp,"%ld\n", (long)getpid());
  fclose(fp);
}

static int
process_commandline(int argc, char **argv, ESL_GETOPTS **ret_go)
{
  ESL_GETOPTS *go = esl_getopts_Create(cmdlineOpts);
  int          status;

  /* if there are no command line arguments, let's try to read /etc/hmmpgmd.conf
   * for any configuration data.
   */
  if (argc == 1) 
    {
      FILE        *fp = NULL;

      if ((fp = fopen(CONF_FILE, "r")) == NULL) 
	{ if (puts("Options --master or --worker must be specified.") < 0) ESL_XEXCEPTION_SYS(eslEWRITE, "write failed"); goto FAILURE; }

      if ((status = esl_opt_ProcessConfigfile(go, CONF_FILE, fp) ) != eslOK)
	{ if (printf("Failed to parse configuration file %s: %s\n",  CONF_FILE, go->errbuf) < 0) ESL_XEXCEPTION_SYS(eslEWRITE, "write failed"); goto FAILURE; }
      fclose(fp);
    } 
  else 
    {
      if (esl_opt_ProcessCmdline(go, argc, argv) != eslOK) 
	{ if (printf("Failed to parse command line: %s\n",  go->errbuf) < 0) ESL_XEXCEPTION_SYS(eslEWRITE, "write failed"); goto FAILURE; }
    }

  if (esl_opt_VerifyConfig(go) != eslOK) { if (printf("Failed to parse command line: %s\n", go->errbuf) < 0) ESL_XEXCEPTION_SYS(eslEWRITE, "write failed"); goto FAILURE; }
 
  /* help format: */
  if (esl_opt_GetBoolean(go, "-h") == TRUE) 
    {
      p7_banner(stdout, argv[0], banner);
      esl_usage(stdout, argv[0], usage);

      if (puts("\nBasic options:") < 0) ESL_XEXCEPTION_SYS(eslEWRITE, "write failed");
      esl_opt_DisplayHelp(stdout, go, 1, 2, 80); /* 1= group; 2 = indentation; 80=textwidth*/

      if (puts("\nOther expert options:") < 0) ESL_XEXCEPTION_SYS(eslEWRITE, "write failed");
      esl_opt_DisplayHelp(stdout, go, 12, 2, 80); 
      exit(0);
    }


  if (esl_opt_ArgNumber(go) != 0) { if (puts("Incorrect number of command line arguments.") < 0) ESL_XEXCEPTION_SYS(eslEWRITE, "write failed"); goto FAILURE; }

  if (esl_opt_IsUsed(go, "--master") && !(esl_opt_IsUsed(go, "--seqdb") || esl_opt_IsUsed(go, "--hmmdb"))) 
    { if (puts("At least one --seqdb or --hmmdb must be specified.") < 0) ESL_XEXCEPTION_SYS(eslEWRITE, "write failed"); goto FAILURE; }

  *ret_go = go;
  return eslOK;
  
 FAILURE:  /* all errors handled here are user errors, so be polite.  */
  esl_usage(stdout, argv[0], usage);
  puts("\nwhere most common options are:");
  esl_opt_DisplayHelp(stdout, go, 1, 2, 80); /* 1= group; 2 = indentation; 80=textwidth*/
  printf("\nTo see more help on available options, do %s -h\n\n", argv[0]);
  esl_getopts_Destroy(go);
  exit(0);  

 ERROR:
  if (go) esl_getopts_Destroy(go);
  exit(status);
}

int
main(int argc, char **argv)
{
  ESL_GETOPTS  *go = NULL;      /* command line processing */

  p7_Init();

  process_commandline(argc, argv, &go);

  /* if we write to a broken socket, ignore the signal and handle the error. */
  our_signal(SIGPIPE, SIG_IGN);

  /* check if we need to write out our pid */
  if (esl_opt_IsOn(go, "--pid")) write_pid(go);

  if      (esl_opt_IsUsed(go, "--master"))  master_process(go);
  else if (esl_opt_IsUsed(go, "--worker"))  worker_process(go);
  else
    { puts("Options --master or --worker must be specified.");  }

  esl_getopts_Destroy(go);

  return eslOK;
}


/*****************************************************************
 * Notes on testing and debugging
 *****************************************************************/


/* How to run a test:
 *   Need a test sequence database: mine is ~eddys/data/hmmpgmd/hmmpgmd.fa
 *   Need a test profile database:  mine is ~eddys/data/hmmpgmd/hmmpgmd.hmm
 *   
 *   src/hmmpgmd --master --seqdb ~/data/hmmpgmd/hmmpgmd.fa --hmmdb ~/data/hmmpgmd/hmmpgmd.hmm &
 *   src/hmmpgmd --worker 127.0.0.1 --cpu 1 &
 *   src/hmmc2 -S
 *     @--seqdb 1
 *     >foo
 *     ACDEFGH...
 *     //
 *     !shutdown
 *     //
 *     
 * Or, for hmmscan against the hmm db, replace @--seqdb 1 with @--hmmdb 1.
 *
 * For debugging, start two of the three processes on cmdline, and the
 * one to be debugged under gdb.
 *    in worker: break process_SearchCmd
 *    
 *    
 * Valgrind debugging
 * generate a small test seq database:
 *     esl-shuffle -G --amino -N 10 -L 400 > foo.fa
 *     fasta2daemon.pl foo.fa              > foo.d
 *     
 * in three separate terminal windows    
 *     valgrind --dsymutil=yes --leak-check=yes --show-reachable=yes src/hmmpgmd --master --seqdb foo.d 
 *     valgrind                --leak-check=yes --show-reachable=yes src/hmmpgmd --worker 127.0.0.1 --cpu 1 
 *     valgrind --dsymutil=yes --leak-check=yes --show-reachable=yes src/hmmc2 -S
 *     
 * copy/paste tutorial/M1.hmm as an example hmmsearch    
 * or any HMM. On Mac OS/X, to "copy" a file my.hmm to the clipboard:
 *    cat my.hmm | pbcopy   
 */


/*****************************************************************
 * @LICENSE@
 *****************************************************************/

